core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -341,9 +341,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
     if(targetSize > originalSize || targetSize < 0)
       throw new KafkaException("Attempt to truncate log segment to " + targetSize + " bytes failed, " +
                                " size of this log segment is " + originalSize + " bytes.")
-    channel.truncate(targetSize)
-    channel.position(targetSize)
-    _size.set(targetSize)
+    if (targetSize != channel.size()) {
+      channel.truncate(targetSize)
+      channel.position(targetSize)
```

Comment:
It seems the code is easier to understand if we always do truncate() and position() together. I agree it would be useful to add a comment to clarify that the expectation is that no other threads will do writes to the log when truncate() is called.